File: Java/MethodInvocationRemoting/src/net/alastairwyse/methodinvocationremoting/ISerializerOperationMap.java
/*
 * Copyright 2013 Alastair Wyse (http://www.oraclepermissiongenerator.net/methodinvocationremoting/)
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *     http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package net.alastairwyse.methodinvocationremoting;

/**
 * Defines methods to store bi-directional mappings between a native Java class type, the type serialized as a string, and the operations to serialize and deserialize objects of that type.
 * @author Alastair Wyse
 */
public interface ISerializerOperationMap {

    /**
     * Adds a mapping.
     * @param nativeType      The native Java type in the mapping.
     * @param serializedType  A string representation of the type.
     * @param serializer      An IObjectSerializer object which performs serialization and deserialization of the type.
     */
    void AddMapping(Class<?> nativeType, String serializedType, IObjectSerializer<?> serializer);
    
    /**
     * Updates (i.e. replaces) an existing mapping.
     * @param nativeType      The native Java type in the mapping.
     * @param serializedType  A string representation of the type.
     * @param serializer      An IObjectSerializer object which performs serialization and deserialization of the type.
     */
    void UpdateMapping(Class<?> nativeType, String serializedType, IObjectSerializer<?> serializer);
    
    /**
     * Retrieves the string representation of a type, based on the inputted native type.
     * @param nativeType  The native type to return the string representation for.
     * @return            The type represented as a string.  Contains null if no mapping entry was found for the native type.
     */
    String GetSerializedType(Class<?> nativeType);
    
    /**
     * Retrieves a native Java type, based on the inputted string representation of the type.
     * @param serializedType  The string representation of the native type.
     * @return                The corresponding native type.  Contains null if no mapping entry was found for the string representation.
     */
    Class<?> GetNativeType(String serializedType);
    
    /**
     * Retrieves the serializer used to serialize and deserialize the type to and from XML.
     * @param nativeType  The native type to return the serializer for.
     * @return            The corresponding serializer.  Contains null if no mapping entry was found for the native type.
     */
    IObjectSerializer<?> GetSerializer(Class<?> nativeType);
    
    /**
     * Retrieves the serializer used to serialize and deserialize the type to and from XML.
     * @param serializedType  The serialized type to return the serializer for.
     * @return                The corresponding serializer.  Contains null if no mapping entry was found for the serialized type.
     */
    IObjectSerializer<?> GetSerializer(String serializedType);
}

File: C#/MethodInvocationRemoting/ISerializerOperationMap.cs
/*
 * Copyright 2013 Alastair Wyse (http://www.oraclepermissiongenerator.net/methodinvocationremoting/)
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *     http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

using System;
using System.Collections.Generic;
using System.Text;
using System.Xml;

namespace MethodInvocationRemoting
{
    /// <summary>
    /// Defines an operation to serialize an object to XML using a System.Xml.XmlWriter.
    /// </summary>
    /// <remarks>This operation should write an XML representation of the object to the XmlWriter, similar to the IXmlSerializable.WriteXml method.</remarks>
    /// <param name="inputObject">The object to serialize.</param>
    /// <param name="writer">The XmlWriter to serialize to.</param>
    public delegate void XmlSerializationOperation(object inputObject, XmlWriter writer);

    /// <summary>
    /// Defines an operation to deserialize an object from XML contained in a System.Xml.XmlReader.
    /// </summary>
    /// <remarks>This operation should generate an object from its XML representation in the XmlReader, similar to the IXmlSerializable.ReadXml method.</remarks>
    /// <param name="reader">The XmlReader to deserialize from.</param>
    /// <returns>The deserialized object.</returns>
    public delegate object XmlDeserializationOperation(XmlReader reader);

    //******************************************************************************
    //
    // Interface: ISerializerOperationMap
    //
    //******************************************************************************
    /// <include file='InterfaceDocumentationComments.xml' path='doc/members/member[@name="T:MethodInvocationRemoting.ISerializerOperationMap"]/*'/>
    public interface ISerializerOperationMap
    {
        /// <include file='InterfaceDocumentationComments.xml' path='doc/members/member[@name="M:MethodInvocationRemoting.ISerializerOperationMap.AddMapping(System.Type,System.String,MethodInvocationRemoting.XmlSerializationOperation,MethodInvocationRemoting.XmlDeserializationOperation)"]/*'/>
        void AddMapping(Type nativeType, string serializedType, XmlSerializationOperation serializationOperation, XmlDeserializationOperation deserializationOperation);

        /// <include file='InterfaceDocumentationComments.xml' path='doc/members/member[@name="M:MethodInvocationRemoting.ISerializerOperationMap.UpdateMapping(System.Type,System.String,MethodInvocationRemoting.XmlSerializationOperation,MethodInvocationRemoting.XmlDeserializationOperation)"]/*'/>
        void UpdateMapping(Type nativeType, string serializedType, XmlSerializationOperation serializationOperation, XmlDeserializationOperation deserializationOperation);

        /// <include file='InterfaceDocumentationComments.xml' path='doc/members/member[@name="M:MethodInvocationRemoting.ISerializerOperationMap.GetSerializedType(System.Type)"]/*'/>
        string GetSerializedType(Type nativeType);

        /// <include file='InterfaceDocumentationComments.xml' path='doc/members/member[@name="M:MethodInvocationRemoting.ISerializerOperationMap.GetNativeType(System.String)"]/*'/>
        Type GetNativeType(string serializedType);

        /// <include file='InterfaceDocumentationComments.xml' path='doc/members/member[@name="M:MethodInvocationRemoting.ISerializerOperationMap.GetSerializationOperation(System.Type)"]/*'/>
        XmlSerializationOperation GetSerializationOperation(Type nativeType);

        /// <include file='InterfaceDocumentationComments.xml' path='doc/members/member[@name="M:MethodInvocationRemoting.ISerializerOperationMap.GetDeserializationOperation(System.String)"]/*'/>
        XmlDeserializationOperation GetDeserializationOperation(string serializedType);
    }
}
